--- ncftp-3.1.5/configure.in.orig	Sun Mar 30 22:22:52 2003
+++ ncftp-3.1.5/configure.in	Sun Mar 30 22:23:53 2003
@@ -138,7 +138,7 @@
 
 AC_CHECK_FUNC(getnameinfo, [
   AC_DEFINE(HAVE_GETNAMEINFO)], [
-  LIBOBJS="$LIBOBJS getnameinfo.o"
+  AC_LIBOBJ([getnameinfo])
   LIBSOBJS="$LIBSOBJS getnameinfo.so"
 ])
 
@@ -233,9 +233,9 @@
   AC_MSG_RESULT(good)
   AC_DEFINE(HAVE_GETADDRINFO)], [
   AC_MSG_RESULT(buggy)
-  LIBOBJS="$LIBOBJS getaddrinfo.o"
+  AC_LIBOBJ([getaddrinfo])
   LIBSOBJS="$LIBSOBJS getaddrinfo.so"])], [
-  LIBOBJS="$LIBOBJS getaddrinfo.o"
+  AC_LIBOBJ([getaddrinfo])
   LIBSOBJS="$LIBSOBJS getaddrinfo.so"
 ])
 
@@ -382,7 +382,6 @@
 AC_SUBST(MAKE)
 AC_SUBST(Z30)
 AC_SUBST(LIBSET)
-AC_SUBST(LIBOBJS)
 AC_SUBST(LIBSOBJS)
 AC_SUBST(MAINDIR)
 AC_SUBST(NCFTP_VERSION)
--- ncftp-3.1.5/autoconf/aclocal.m4.orig	Sun Oct 13 07:03:49 2002
+++ ncftp-3.1.5/autoconf/aclocal.m4	Sun Mar 30 22:43:14 2003
@@ -2515,27 +2515,26 @@
 if test -f "$h_file" && $test_not_L "$h_file" ; then
 	h_tmp=`echo "$h_file" | sed 's/\.h/.tmp/;'`
 
-changequote(<<, >>)dnl
 	remove_cpp_warning=yes
-	if [ "$SYS" = linux ] ; then
+	if test "$SYS" = linux  ; then
 		#
 		# We need to retain #warning on Linux
 		#
 		remove_cpp_warning=no
 	fi
-	if [ "$ac_cv_prog_gcc" = yes ] ; then
+	if test "$ac_cv_prog_gcc" = yes  ; then
 		#
 		# GCC accepts #warning
 		#
 		remove_cpp_warning=no
 	fi
 
-	if [ "$remove_cpp_warning" = yes ] ; then
+	if test "$remove_cpp_warning" = yes  ; then
 		#
 		# Remove "#warning", since compiler will complain about it
 		# not being recognized.
 		#
-		sed 's/^\([\ \	]*#[\ \	]*warning.*\)/\/* \1 *\//;' "$h_file" > "$h_tmp"
+		sed ['s/^\([\ \	]*#[\ \	]*warning.*\)/\/* \1 *\//;'] "$h_file" > "$h_tmp"
 		cmp -s "$h_file" "$h_tmp"
 		if test $? -ne 0 ; then 
 			mv "$h_tmp" "$h_file"
@@ -2544,7 +2543,6 @@
 		/bin/rm -f "$h_tmp"
 		unset h_tmp longest_int_subst
 	fi
-changequote([, ])dnl
 fi
 unset h_file remove_cpp_warning
 ])
