--- ncftp-3.2.5/configure.in.orig	2011-03-27 15:57:04.305737273 +0200
+++ ncftp-3.2.5/configure.in	2011-03-27 16:00:02.729737273 +0200
@@ -42,7 +42,7 @@
 
 dnl Try to use PATH rather than hardcode the installation path, if possible.
 if test "${prefix-NONE}" != "NONE" && test "$prefix" != "/usr/local" && test "$prefix" != "/usr"; then
-	AC_DEFINE_UNQUOTED(PREFIX_BINDIR, "$prefix/bin")
+	AC_DEFINE_UNQUOTED(PREFIX_BINDIR, "$prefix/bin", [bin directory])
 fi
 
 
@@ -152,7 +152,7 @@
        ipv6=no
        ;;
   *)   AC_MSG_RESULT(yes)
-       AC_DEFINE(ENABLE_IPV6)
+       AC_DEFINE(ENABLE_IPV6,,[Define if you want IPv6 support.])
        ipv6=yes
        ;;
   esac ],
@@ -179,13 +179,13 @@
 ))
 
 AC_CHECK_FUNC(getaddrinfo, [
-  AC_DEFINE(HAVE_GETADDRINFO)], [
+  AC_DEFINE(HAVE_GETADDRINFO,,[Define if you have the getaddrinfo function.])], [
   AC_LIBOBJ([getaddrinfo])
   LIBSOBJS="$LIBSOBJS getaddrinfo.so"
 ])
 
 AC_CHECK_FUNC(getnameinfo, [
-  AC_DEFINE(HAVE_GETNAMEINFO)], [
+  AC_DEFINE(HAVE_GETNAMEINFO,,[Define if you have the getaddrname function.])], [
   AC_LIBOBJ([getnameinfo])
   LIBSOBJS="$LIBSOBJS getnameinfo.so"
 ])
@@ -197,7 +197,7 @@
 [struct sockaddr x;
 x.sa_len = 0;],
        AC_MSG_RESULT(yes)
-       AC_DEFINE(HAVE_SOCKADDR_SA_LEN),
+       AC_DEFINE(HAVE_SOCKADDR_SA_LEN,,[Define if sockaddr has sa_len member]),
        AC_MSG_RESULT(no))
 
 # check if sockaddr_storage has ss_family member
@@ -207,9 +207,9 @@
 [struct sockaddr_storage x;
 x.__ss_family = 0;],
 	AC_MSG_RESULT(yes)
-	AC_DEFINE(HAVE_SOCKADDR_STORAGE_SS_FAMILY),
+	AC_DEFINE(HAVE_SOCKADDR_STORAGE_SS_FAMILY,,[Define if sockaddr_storage has ss_family member]),
 	AC_MSG_RESULT(no)
-	AC_DEFINE(HAVE_SOCKADDR_STORAGE___SS_FAMILY)
+	AC_DEFINE(HAVE_SOCKADDR_STORAGE___SS_FAMILY,,[Define if sockaddr_storage has __ss_family member])
 	CPPFLAGS="$CPPFLAGS -D__ss_family=ss_family -D__ss_len=ss_len")
 
 CPPFLAGS="-I$MAINDIR $CPPFLAGS"
