diff -Nur ncftp-3.0.2.orig/ncftp/syshdrs.h ncftp-3.0.2/ncftp/syshdrs.h
--- ncftp-3.0.2.orig/libncftp/syshdrs.h	Wed Oct 18 02:25:50 2000
+++ ncftp-3.0.2/libncftp/syshdrs.h	Wed Apr  4 23:40:23 2001
@@ -183,6 +183,12 @@
 #	define Lseek(a,b,c) lseek(a, (off_t) b, c)
 #endif
 
+#if !defined(HAVE_SOCKADDR_SA_LEN) && !defined(SA_LEN)
+#	define SA_LEN(x) ( \
+		((x)->sa_family == AF_INET6) ? sizeof(struct sockaddr_in6) : \
+		(((x)->sa_family == AF_INET) ? sizeof(struct sockaddr_in) : \
+		sizeof(struct sockaddr)))
+#endif
 
 #include <Strn.h>			/* Library header. */
 #include <ncftp.h>			/* Library header. */
diff -Nur ncftp-3.0.2.orig/sio/syshdrs.h ncftp-3.0.2/sio/syshdrs.h
--- ncftp-3.0.2.orig/sio/syshdrs.h	Wed Oct 18 02:26:38 2000
+++ ncftp-3.0.2/sio/syshdrs.h	Wed Apr  4 23:40:56 2001
@@ -189,6 +189,13 @@
 #define SYNCH   242             /* for telfunc calls */
 #endif
 
+#if !defined(HAVE_SOCKADDR_SA_LEN) && !defined(SA_LEN)
+#	define SA_LEN(x) ( \
+		((x)->sa_family == AF_INET6) ? sizeof(struct sockaddr_in6) : \
+		(((x)->sa_family == AF_INET) ? sizeof(struct sockaddr_in) : \
+		sizeof(struct sockaddr)))
+#endif
+
 #include "sio.h"			/* Library header. */
 #ifndef NO_UNIX_DOMAIN_SOCKETS
 #	include "usio.h"
