--- ncftp-3.1.7/sio/SAccept.c.orig	2004-01-08 21:54:31.000000000 +0100
+++ ncftp-3.1.7/sio/SAccept.c	2004-01-08 22:01:33.093223328 +0100
@@ -3,6 +3,18 @@
 #	pragma hdrstop
 #endif
 
+#ifndef HAVE_SOCKADDR_SA_LEN
+#ifndef SA_LEN
+#ifdef ENABLE_IPV6
+#define SA_LEN(x) (((x)->sa_family == AF_INET6) ? sizeof(struct sockaddr_in6) \
+		: (((x)->sa_family == AF_INET) ? sizeof(struct sockaddr_in) \
+		: -1))
+#else  /* ENABLE_IPV6 */
+#define SA_LEN(x) (((x)->sa_family == AF_INET) ? sizeof(struct sockaddr_in) : -1)
+#endif /* ENABLE_IPV6 */
+#endif /* SA_LEN */
+#endif /* HAVE_SOCKADDR_SA_LEN */
+
 int
 SAccept(int sfd, struct sockaddr *const addr, int tlen)
 {
