--- ncftp-3.2.2/libncftp/Makefile.in.orig	2008-08-19 06:44:17.000000000 +0200
+++ ncftp-3.2.2/libncftp/Makefile.in	2008-08-27 19:38:09.914117548 +0200
@@ -6,6 +6,7 @@
 PKGNAME=libncftp
 VERSION=@LIBNCFTP_VERSION@
 PREFIX=@prefix@
+LIBDIR=@libdir@
 
 @SET_MAKE@
 CC=@CC@
@@ -65,24 +66,24 @@
 	-@/bin/ls -l $(LIBSOS) $(LIBSO)
 
 soinstall: $(LIBSO)
-	[ -d "$(PREFIX)/lib" ] || mkdir -m 755 "$(PREFIX)/lib"
-	[ -d "$(PREFIX)/include" ] || mkdir -m 755 "$(PREFIX)/include"
-	cp $(LIBSO) $(PREFIX)/lib/$(LIBSO)
-	( cd $(PREFIX)/lib ; /bin/rm -f $(LIBSOS) ; /bin/ln -s $(LIBSO) $(LIBSOS) )
-	cp ncftp.h ncftp_errno.h $(PREFIX)/include
-	-chmod a+r $(PREFIX)/lib/$(LIBSO) $(PREFIX)/include/ncftp.h $(PREFIX)/include/ncftp_errno.h
-	-@/bin/ls -ld $(PREFIX)/lib/$(LIBSO)
+	[ -d "$(DESTDIR)$(LIBDIR)" ] || mkdir -m 755 "$(DESTDIR)$(LIBDIR)"
+	[ -d "$(DESTDIR)$(PREFIX)/include" ] || mkdir -m 755 "$(DESTDIR)$(PREFIX)/include"
+	cp $(LIBSO) $(DESTDIR)$(LIBDIR)/$(LIBSO)
+	( cd $(DESTDIR)$(LIBDIR) ; /bin/rm -f $(LIBSOS) ; /bin/ln -s $(LIBSO) $(LIBSOS) )
+	cp ncftp.h ncftp_errno.h $(DESTDIR)$(PREFIX)/include
+	-chmod a+r $(DESTDIR)$(LIBDIR)/$(LIBSO) $(DESTDIR)$(PREFIX)/include/ncftp.h $(DESTDIR)$(PREFIX)/include/ncftp_errno.h
+	-@/bin/ls -ld $(DESTDIR)$(LIBDIR)/$(LIBSO)
 	-@echo
 	-@echo "Make sure that your dynamic loader searches for the directory $(PREFIX)/lib."
 	-@echo "For example, edit /etc/ld.so.conf accordingly if your system uses ld.so."
 
 install: $(LIB)
-	[ -d "$(PREFIX)/lib" ] || mkdir -m 755 "$(PREFIX)/lib"
-	[ -d "$(PREFIX)/include" ] || mkdir -m 755 "$(PREFIX)/include"
-	cp $(LIB) $(PREFIX)/lib/$(LIB)
-	-@@RANLIB@ $(PREFIX)/lib/$(LIB)
-	cp ncftp.h ncftp_errno.h $(PREFIX)/include
-	-chmod a+r $(LIB) $(PREFIX)/lib/$(LIB) $(PREFIX)/include/ncftp.h $(PREFIX)/include/ncftp_errno.h
+	[ -d "$(DESTDIR)$(LIBDIR)" ] || mkdir -m 755 "$(DESTDIR)$(LIBDIR)"
+	[ -d "$(DESTDIR)$(PREFIX)/include" ] || mkdir -m 755 "$(DESTDIR)$(PREFIX)/include"
+	cp $(LIB) $(DESTDIR)$(LIBDIR)/$(LIB)
+	-@@RANLIB@ $(DESTDIR)$(LIBDIR)/$(LIB)
+	cp ncftp.h ncftp_errno.h $(DESTDIR)$(PREFIX)/include
+	-chmod a+r $(LIB) $(DESTDIR)$(LIBDIR)/$(LIB) $(DESTDIR)$(PREFIX)/include/ncftp.h $(DESTDIR)$(PREFIX)/include/ncftp_errno.h
 	( cd ../sio ; $(MAKE) "PREFIX=$(PREFIX)" install)
 	( cd ../Strn ; $(MAKE) "PREFIX=$(PREFIX)" install_min )
 
