--- ncftp-3.1.5/configure.in.orig	Sun Mar 30 22:22:52 2003
+++ ncftp-3.1.5/configure.in	Sun Mar 30 22:23:53 2003
@@ -138,7 +138,7 @@
 
 AC_CHECK_FUNC(getnameinfo, [
   AC_DEFINE(HAVE_GETNAMEINFO)], [
-  LIBOBJS="$LIBOBJS getnameinfo.o"
+  AC_LIBOBJ([getnameinfo])
   LIBSOBJS="$LIBSOBJS getnameinfo.so"
 ])
 
@@ -233,9 +233,9 @@
   AC_MSG_RESULT(good)
   AC_DEFINE(HAVE_GETADDRINFO)], [
   AC_MSG_RESULT(buggy)
-  LIBOBJS="$LIBOBJS getaddrinfo.o"
+  AC_LIBOBJ([getaddrinfo])
   LIBSOBJS="$LIBSOBJS getaddrinfo.so"])], [
-  LIBOBJS="$LIBOBJS getaddrinfo.o"
+  AC_LIBOBJ([getaddrinfo])
   LIBSOBJS="$LIBSOBJS getaddrinfo.so"
 ])
 
@@ -382,7 +382,6 @@
 AC_SUBST(MAKE)
 AC_SUBST(Z30)
 AC_SUBST(LIBSET)
-AC_SUBST(LIBOBJS)
 AC_SUBST(LIBSOBJS)
 AC_SUBST(MAINDIR)
 AC_SUBST(NCFTP_VERSION)
