--- ncftp-3.2.1/configure.in.orig	2007-08-12 13:27:12.609525915 +0200
+++ ncftp-3.2.1/configure.in	2007-08-12 13:40:36.431333110 +0200
@@ -146,13 +146,13 @@
 
 AC_CHECK_FUNC(getaddrinfo, [
   AC_DEFINE(HAVE_GETADDRINFO)], [
-  LIBOBJS="$LIBOBJS getaddrinfo.o"
+  AC_LIBOBJ([getaddrinfo])
   LIBSOBJS="$LIBSOBJS getaddrinfo.so"
 ])
 
 AC_CHECK_FUNC(getnameinfo, [
   AC_DEFINE(HAVE_GETNAMEINFO)], [
-  LIBOBJS="$LIBOBJS getnameinfo.o"
+  AC_LIBOBJ([getnameinfo])
   LIBSOBJS="$LIBSOBJS getnameinfo.so"
 ])
 
@@ -323,7 +323,6 @@
 AC_SUBST(Z30)
 AC_SUBST(Z31)
 AC_SUBST(LIBSET)
-AC_SUBST(LIBOBJS)
 AC_SUBST(LIBSOBJS)
 AC_SUBST(MAINDIR)
 AC_SUBST(NCFTP_VERSION)
